### ISSUE TITLE
Large delay between batches when training

### ISSUE THREAD

**Issue by abagshaw** [AlexeyAB/darknet#585]
I'm training on a GCP instance with a Tesla P100, 4vCPUs, 12GB RAM and 100GB SSD.

I'm noticing that between batches there is a large delay before the next batch starts - around 10 seconds (i.e. after avg loss is printed and time taken to load data is printed):

Given that performance of data load is already monitored and printed out (ranges from 1s to 2s in my case) I'm assuming that the extra 8ish seconds are due to something in this block:

https://github.com/AlexeyAB/darknet/blob/master/src/classifier.c#L134-L145

I'm guessing the culprit is ([https://github.com/AlexeyAB/darknet/blob/master/src/classifier.c#L134](`free_data(train);`)) but I wouldn't expect that freeing up memory should take 8 seconds. Am I missing something, or is there some other way to optimize this/remove the latency?

Thanks very much!

**Comment by AlexeyAB** [comment]
As I see you train train the classifier on GCP dataset. There is **bottleneck** in data loading: https://github.com/AlexeyAB/darknet/blob/master/src/classifier.c#L114-L116
    More specifically in **data augmentation**: https://github.com/AlexeyAB/darknet/blob/2a9f7e44ce1b73d3d56ef83f83e94f074ecac3f9/src/data.c#L107-L132



I accelerated data augmentation 3.5x times only for training detector, but didn't it for classifier.
If I will have a time, I will increase speed of data augmentation for classifier too.

**Comment by abagshaw** [comment]
@AlexeyAB Yes, I am loading from a GCP Storage Bucket (so I understand that data loading latency will be increased - compared to reading from a local drive).

However, as I mentioned, the latency in data loading appears to be already measured (and printed [here](https://github.com/AlexeyAB/darknet/blob/master/src/classifier.c#L118)) The loading latency as measured there is consistently 1 - 2 seconds for me, however there is an **additional** 8 sec approx latency between completion of the previous batch and printing of loading latency.

I can only see two potential explanations:
1. Something is wrong with the time measurement for data loading (data loading is taking longer than is being reported) - unlikely
2. The bottleneck is somewhere (here)[https://github.com/AlexeyAB/darknet/blob/master/src/classifier.c#L134-L145]. The only place I could see a bottleneck happening in that block is in [`free_data(data)`](https://github.com/AlexeyAB/darknet/blob/master/src/classifier.c#L134).

In summary, I don't think the problem is simply data loading latency - but a bottleneck somewhere else.

**Comment by AlexeyAB** [comment]
Just try to keep only 1 image in the `train.txt` file that located on SSD-drive and check train.
`Loaded: 2.0000 secconds`  - isn't the time that spent for loading data. 

1. If you train on Linux, then `clock()` shows CPU-time instead of real-time, so it can show incorrect value of time (higher if many cores are occupied, lower if threads sleep or waits for the CUDA-kernel function to finish)

2. Look at this code, the message `Loaded: 2.0000 secconds` - indicates the value of the time that the data load was performed longer than network training - i.e. `2.0 sec = load_time - network_time`:


```cpp
        time=clock();

        pthread_join(load_thread, 0);  // wait for finishing load-thread
        train = buffer;
        load_thread = load_data(args); // run new load-thread

        printf("Loaded: %lf seconds\n", sec(clock()-time));
        time=clock();

        // here is non-measured time when occur Load_data and Train_network in parallel
```


https://github.com/AlexeyAB/darknet/blob/2a9f7e44ce1b73d3d56ef83f83e94f074ecac3f9/src/classifier.c#L112-L119

**Comment by Ashwin-Ramesh2607** [comment]
@abagshaw You have mentioned that you are using darknet to directly load data from GCS bucket, could you direct me on how to proceed with that?
